### ISSUE TITLE
How many epochs are needed to train on LJSpeech dataset from scratch?

### ISSUE THREAD

**Issue by chenming6615** [NVIDIA/tacotron2#522]
I have trained the model from scratch on the  LJSpeech dataset with batch_size of 5 on a GTX 1650 GPU. 
The validation loss seems converged around 0.6, after about 10 hours, 16000 iterations. 
I  ran inference with inference.ipynb  replacing `tacotron2_statedict.pt ` with the generated ` outdir/checkpoint_16000` file.

However, the generated .wav file is still awful.  It's almost pure noise. 

Is it because the training time is  not enough or because the batch_size is two small?

I did change any code and hyper-parameter except the batch_size.

**Comment by CookiePPP** [comment]
> Is it because the training time is not enough or because the batch_size is two small?

Both actually.
Learning rate is roughly proportional to the square-root of the batch-size.

---

Using this paper as reference;
https://arxiv.org/pdf/1910.10288.pdf

You'll need 
`0.001 * (5/256)**0.5` = `0.00014` or **`1.4e-4` learning rate** if you use batch_size = 5

(256 = batch size of the paper)
(0.001 = learning rate of the paper)
(5 = batch size you are using)
( `**0.5` = applying square root )

---

![image](https://user-images.githubusercontent.com/42448678/132212784-2c1356b5-95d5-445a-a8af-da215f3da572.png)

The paper used 256 batch size, and tacotron2 uses Location-Sensitive Attention.
The model started understanding the text between 1500 and 3000 iterations at batch_size 256 as that's when the error values in the graph start dropping.

Their batch size is 51 times larger than yours, so you need between 76,800 and 153,600 iterations for the model to start working.

**Comment by CookiePPP** [comment]
I recommend using Google Colab for free faster GPUs, and using [pretrained models](https://github.com/NVIDIA/tacotron2#training-using-a-pre-trained-model) to reduce the amount of iterations required for good results.

**Comment by chenming6615** [comment]
@CookiePPP Thanks for your kind reply.   Now I am training it on a RTX 3090 GPU, with batch_size of 80. 
After 21709 iterations, with the validation loss converging around 0.35. 
Now, it sounds like a human , but  still, she is like speaking a totally different language other than English.

**Comment by chenming6615** [comment]
> and using [pretrained models](https://github.com/NVIDIA/tacotron2#training-using-a-pre-trained-model) to reduce the amount of iterations required for good results.

I am trying to reproduce the result to proof that this algorithm really works.

**Comment by CookiePPP** [comment]
@chenming6615 Did it end up working?

**Comment by lutakrystal305** [comment]
@CookiePPP Sr, can you tell me which dataset is fit to train difference language? tks :D

**Comment by lutakrystal305** [comment]
hour, from one or multi people, not noise?

**Comment by CookiePPP** [comment]
@lutakrystal305 
I do not know.

---

> hour, from one or multi people, not noise?

![image](https://user-images.githubusercontent.com/42448678/132938901-9e11cf0a-b457-45e6-a4a7-4d36c16f70ce.png)

1 Person. 4 Hours of data minimum. No background noise should be allowed.

**Comment by ksaidin** [comment]
> @CookiePPP Thanks for your kind reply. Now I am training it on a RTX 3090 GPU, with batch_size of 80.
> After 21709 iterations, with the validation loss converging around 0.35.
> Now, it sounds like a human , but still, she is like speaking a totally different language other than English.

The fact you can run this repo under rtx3090 is interesting. can you please explain how you've managed to install tensorflow 1.15 while you are forced to install Driver 450 (which leads to CUDA 11.0, therefore cudnn 8.0) consequently the least version of tensorflow you could install is tf 2.5?

**Comment by theaimeltus** [comment]
@ksaidin Well, I used an RTX 3050, and mine worked as well! From my experiences using my RTX 3050 to train using this repo, I used Miniconda! I used pip to force install Tensorflow 1.15.2 on it (pip install tensorflow==1.15.2), I also have to use the requirement.txt's specific version of librosa and additionally, numbo. For the rest of the requirements.txt, I just install the latest versions on it... 
To get PyTorch working on it, I must install CUDA 11.1 Update 1 specifically as firstly, it supports RTX devices with it and secondly, it's one of the only versions that you can install from PyTorch's website. I installed both PyTorch and Apex using conda (conda install nvidia-apex) (conda install conda install pytorch torchvision torchaudio cudatoolkit=11.1 -c pytorch -c conda-forge).
Finally, to get it training properly with coherent pronunciations of words, I've used @CookiePPP 's formula of "0.001 * (batch size/256)**0.5" to get a suitable learning rate, so I could use the batch size of 4 (since RTX 3050 only has a 4GB VRAM, not suited I know, but I got it to work!). 
Initially at my first test, I used the preset default learning rate of 1e-3 from NVidia. It was able to train my model, but somehow the pronunciations got so messed up that it doesn't sound like the speaker at all. But after I used Cookie's learning rate formula and trained from it, my model was able to pronounce things as it should be, also sounding almost accurately like the speaker. 
So, Cookie's learning rate formula was able to help me train my model greatly and I thank them for the advice! Anyhow, that's how I was able to train using this repo with an RTX 3050.

**Comment by chenming6615** [comment]
> @ksaidin Well, I used an RTX 3050, and mine worked as well! From my experiences using my RTX 3050 to train using this repo, I used Miniconda! I used pip to force install Tensorflow 1.15.2 on it (pip install tensorflow==1.15.2), I also have to use the requirement.txt's specific version of librosa and additionally, numbo. For the rest of the requirements.txt, I just install the latest versions on it...
> To get PyTorch working on it, I must install CUDA 11.1 Update 1 specifically as firstly, it supports RTX devices with it and secondly, it's one of the only versions that you can install from PyTorch's website. I installed both PyTorch and Apex using conda (conda install nvidia-apex) (conda install conda install pytorch torchvision torchaudio cudatoolkit=11.1 -c pytorch -c conda-forge).
> Finally, to get it training properly with coherent pronunciations of words, I've used @CookiePPP 's formula of "0.001 * (batch size/256)**0.5" to get a suitable learning rate, so I could use the batch size of 4 (since RTX 3050 only has a 4GB VRAM, not suited I know, but I got it to work!).
> Initially at my first test, I used the preset default learning rate of 1e-3 from NVidia. It was able to train my model, but somehow the pronunciations got so messed up that it doesn't sound like the speaker at all. But after I used Cookie's learning rate formula and trained from it, my model was able to pronounce things as it should be, also sounding almost accurately like the speaker.
> So, Cookie's learning rate formula was able to help me train my model greatly and I thank them for the advice! Anyhow, that's how I was able to train using this repo with an RTX 3050.

Thanks for your reply!  Would you please tell us how many hours it is needed to train it with just a RTX 3050 ?

**Comment by theaimeltus** [comment]
> Thanks for your reply! Would you please tell us how many hours it is needed to train it with just a RTX 3050 ?

@chenming6615 Well, it really depends on how big is the dataset, but since the dataset I trained on is quite small, it took me about 9-14 hours to get a decent voice out of it. So, if I would to have a bigger dataset (like LJSpeech), it would probably take me about a day or two (the most would be three days)... It may sound like it's very long, but for me, it's actually quite such a small time compared to other stuff I've seen! (I also forgot to mention that I trained the dataset on the pretrained model provided by this repo)

From my experiences and talking with others, when training with a pretrained model, it is recommended to stop training once it got under 0.1 validation loss (which is what I've done).

**Comment by lutakrystal305** [comment]
@CookiePPP can you help me?
This model can't converge and inference will incorrect?
![image](https://user-images.githubusercontent.com/65547732/134756180-693c7e35-183e-45b7-9bc0-062a1af10fef.png) // This is alignment inference from my model.
![image](https://user-images.githubusercontent.com/65547732/134756217-80b24dbc-8ec6-4ad7-854b-d3d5c47b72a8.png)
//This is my tensorboard at step 34k.
![image](https://user-images.githubusercontent.com/65547732/134756239-272d33b3-d52d-45d9-8bfb-168cca680cef.png)
//This is validate and training loss
tks you :D

**Comment by lutakrystal305** [comment]
If i train a different language, should i train from scratch or pretrain tacotron2?

**Comment by CookiePPP** [comment]
> This model can't converge and inference will incorrect?

Yes, that model is probably never going to learn. Once the alignment gets stuck near the top (end of sentence), it normally never unlearns that.

You will need to restart training.
Using pretrained tacotron2 is still a good idea, but before you do another training run make sure;

- all audio files do not have long stretches of silence
- there is no silence at the start/end (trim the audio files if you haven't already)
- the text doesn't have any constant characters used in the same place every time. e.g: If every text ends in `.`, the model might start thinking the `.` at the end of the sentence always means silence anywhere in the audio file.

If you can't get it working with these changes then look into guiding the attention.

also, I've already written a guide that might be helpful for you.
https://github.com/NVIDIA/tacotron2/issues/321#issuecomment-603894212

**Comment by BlazJurisic** [comment]
@CookiePPP, any recommendations for a weight_decay hparam?

**Comment by CookiePPP** [comment]
@BlazJurisic No recommendations from me. I haven't experimented with that parameter before

**Comment by sabat84** [comment]
Hi, I wanted to train tacotron 2 from scratch with 4652 sentences (Kurdish dataset) (10 hours), batch size 32. here are some plots:
![Capture](https://user-images.githubusercontent.com/103031919/161809103-1a6c2f39-7d8a-4e9e-9cf8-0b5174e9d5fa.JPG)
![Capture1](https://user-images.githubusercontent.com/103031919/161809679-75be72e3-55bd-435f-91d1-a9bb4cfe2e85.JPG)
does this model are going to be converge or not? please help @CookiePPP

**Comment by CookiePPP** [comment]
@sabat84 
This looks good. It should converge given more time.

**Comment by sabat84** [comment]
@CookiePPP 
How many steps I have to wait?

**Comment by CypherpunkSamurai** [comment]
> * the text doesn't have any constant characters used in the same place every time. e.g: If every text ends in ., the model might start thinking the . at the end of the sentence always means silence anywhere in the audio file.

Great Point there. Also, can i assume using `,` and `-` can help punctuation?
